File: java/leetcode/src/Task114.java
import org.junit.jupiter.api.Test;

public class Task114 {
    public void flatten(TreeNode root) {
        if(root == null){
            return;
        }

        recursion2(root);

    }

    TreeNode recursion(TreeNode root){
        if(root.left == null && root.right == null){
            return root;
        }

        TreeNode tail = root;
        if(root.left != null){
            TreeNode tailL = recursion(root.left);
            tailL.right = root.right;
            root.right = root.left;
            root.left = null;
            tail = tailL;
        }

        if(tail.right != null){
            tail = recursion(tail.right);
        }

        return tail;
    }

    TreeNode recursion2(TreeNode root){
        if(root == null ){
            return root;
        }

        TreeNode tail = root;
        TreeNode tailL = recursion(root.left);
        TreeNode tailR = recursion(root.right);

        if(tailL != null){
            tail = tailL;
            tailL.right = root.right;
            root.right = root.left;
            root.left = null;
        }

        if(tailR!= null){
            tail = tailR;
        }

        return tail;
    }

    @Test
    void test(){
        TreeNode root = new TreeNode(1).left(2).right(5);
        root.left.left(3).right(4);
        root.right.right(6);
        System.out.println(root);
        flatten(root);
        System.out.println(root);
    }

}

File: py/Task114.py

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # dfs using recursion
    def flatten(self, root: TreeNode) -> None:
        if not root:
            return

        def flatten(root: TreeNode) -> TreeNode:
            tail = root
            # left
            if root.left:
                tail = flatten(root.left)
                root.left, root.right, tail.right = None, root.left, root.right
            # right
            if tail.right:
                tail = flatten(tail.right)
            return tail
        flatten(root)


obj = Solution()
root = TreeNode(1)
root.left, root.right = TreeNode(2), TreeNode(5)
root.left.left, root.left.right = TreeNode(3), TreeNode(4)
root.right.right = TreeNode(6)
obj.flatten(root)
